app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts

```diff
@@ -0,0 +1,61 @@
+import { Application } from 'core/application/application.model';
+import { AccountService } from './../../account/account.service';
+import { IModalService } from 'angular-ui-bootstrap';
+import { IPromise, module, IQService, auto } from 'angular';
+
+import { ACCOUNT_SERVICE } from 'core/account/account.service';
+import { CLOUD_PROVIDER_REGISTRY, CloudProviderRegistry } from 'core/cloudProvider/cloudProvider.registry';
+import { SETTINGS } from 'core/config/settings';
+
+export class ProviderSelectionService {
+  constructor(private $uibModal: IModalService,
+              private $q: IQService,
+              private accountService: AccountService,
+              private cloudProviderRegistry: CloudProviderRegistry) {
+    'ngInject';
+  }
+
+  public selectProvider(application: Application, feature: string): IPromise<string> {
+    return this.accountService.listProviders(application).then((providers) => {
+      let provider;
+      let reducedProviders: string[] = [];
+      if (feature) {
+        reducedProviders = providers.filter((p) => this.cloudProviderRegistry.hasValue(p, feature));
+      }
+
+      // reduce the providers the smallest, unique collection taking into consideration the useProvider values
+      reducedProviders = [...new Set(reducedProviders.map((_provider) => {
+        let result = _provider;
+        const providerFeature = this.cloudProviderRegistry.getProvider(_provider)[feature];
+        if (providerFeature) {
+          result = providerFeature.useProvider || _provider;
+        }
+        return result;
+      }))];
+
+      if (reducedProviders.length > 1) {
+        provider = this.$uibModal.open({
+          templateUrl: require('./providerSelection.html'),
+          controller: 'ProviderSelectCtrl as ctrl',
+          resolve: {
+            providerOptions: function() { return reducedProviders; }
+          }
+        }).result;
+      } else if (reducedProviders.length === 1) {
+        provider = this.$q.when(reducedProviders[0]);
+      } else {
+        provider = this.$q.when(SETTINGS.defaultProvider || 'aws');
+      }
+      return provider;
+    });
+  }
+}
+
+export let providerSelectionService: ProviderSelectionService;
+export const PROVIDER_SELECTION_SERVICE = 'spinnaker.cloudProvider.providerSelection.service';
+module (PROVIDER_SELECTION_SERVICE, [
+  ACCOUNT_SERVICE,
+  CLOUD_PROVIDER_REGISTRY,
+]).factory('providerSelectionService', ($uibModal: IModalService, $q: IQService, accountService: AccountService, cloudProviderRegistry: CloudProviderRegistry) =>
```

Comment:
Any reason to use a factory over a service here? Also, the way you've invoked the class, I don't think you need `'ngInject'` above.